common/src/main/kotlin/streams/extensions/CommonExtensions.kt

```diff
@@ -34,9 +39,45 @@ fun Map<String, Any?>.flatten(map: Map<String, Any?> = this, prefix: String = ""
     }.toMap()
 }
 
-fun <K, V> ConsumerRecord<K, V>.topicPartition() = TopicPartition(this.topic(), this.partition())
-fun <K, V> ConsumerRecord<K, V>.offsetAndMetadata(metadata: String = "") = OffsetAndMetadata(this.offset() + 1, metadata)
+fun ConsumerRecord<*, *>.topicPartition() = TopicPartition(this.topic(), this.partition())
+fun ConsumerRecord<*, *>.offsetAndMetadata(metadata: String = "") = OffsetAndMetadata(this.offset() + 1, metadata)
 
-fun ConsumerRecord<ByteArray, ByteArray>.toStreamsSinkEntity(): StreamsSinkEntity = StreamsSinkEntity(
-        if (this.key() == null) null else JSONUtils.readValue<Any>(this.key(), true),
-        if (this.value() == null) null else JSONUtils.readValue<Any>(this.value()))
\ No newline at end of file
+private fun convertAvroData(rawValue: Any?): Any? = when (rawValue) {
+    is IndexedRecord -> rawValue.toMap()
+    is Collection<*> -> rawValue.toList().map { convertAvroData(it) }
+    is Map<*, *> -> rawValue
+            .mapKeys { it.key.toString() }
+            .mapValues { convertAvroData(it.value) }
+    is GenericFixed -> rawValue.bytes()
+    is ByteBuffer -> rawValue.array()
+    is GenericEnumSymbol, is CharSequence -> rawValue.toString()
+    else -> rawValue
+}
+fun IndexedRecord.toMap() = this.schema.fields
+        .map { it.name() to convertAvroData(this[it.pos()]) }
+        .toMap()
+
+fun Schema.toMap() = JSONUtils.asMap(this.toString())
+
+private fun convertData(data: Any?): Any? {
+    return data?.let {
+        when (data) {
+            is ByteArray -> JSONUtils.readValue<Any>(data)
+            is GenericRecord -> data.toMap()
+            else -> throw RuntimeException("Unsupported type ${data::class.java.name}")
+        }
+    }
+}
+fun ConsumerRecord<*, *>.toStreamsSinkEntity(): StreamsSinkEntity {
+    val rawKey = this.key()
+    val key = try {
+        convertData(rawKey)
+    } catch(e: Exception) {
```

Comment:
is Exception here too generic, what about that runtime exception from above about the unsupported type? should we turn that value really into a string ? or fail?

Reply:
what about something like this?

```Kotlin
private fun convertData(data: Any?, stringWhenFailure: Boolean = false): Any? {
    return when (data) {
        is ByteArray -> JSONUtils.readValue<Any>(data)
        is GenericRecord -> data.toMap()
        null -> null
        else -> if (stringWhenFailure) data.toString() else throw RuntimeException("Unsupported type ${data::class.java.name}")
    }
}
fun ConsumerRecord<*, *>.toStreamsSinkEntity(): StreamsSinkEntity {
    val key = convertData(this.key(), true)
    val value = convertData(this.value())
    return StreamsSinkEntity(key, value)
}
```

Reply:
yep